portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -46,46 +49,66 @@ public void init(){
     Query query = new Query("Comment").addSort("date", SortDirection.DESCENDING);
 
+    UserService userService = UserServiceFactory.getUserService();
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
+    if (userService.isUserLoggedIn()) {
+        List<Comment> comments = new ArrayList<>();
 
-    List<String> comments = new ArrayList<>();
-    if (!set) { quantity = 10; }  // Default set at max of 10 comments shown upon loading screen
-    int i = 0;
-    for (Entity entity : results.asIterable()) {
-      // Limits number of comments added to the page
-      if (i < quantity){ 
-        long id = entity.getKey().getId();
-        String message = (String) entity.getProperty("comment");
-        Date timestamp = (Date) entity.getProperty("date");
-        comments.add(new Comment(id, message, timestamp));
-        i++;
-      } else {
-        break; // Exits for-loop once requested number of comments appear
-      }
-    }
+        // Add the current user to the JSON to send over to the JS file
+        comments.add(new Comment(0, null, null, userService.getCurrentUser().getEmail()));
 
-    response.setContentType("application/json");
-    String json = new Gson().toJson(comments);
-    response.getWriter().println(json);
+        if (!set) { quantity = 10; }  // Default set at max of 10 comments shown upon loading screen
+        int i = 0;
+        for (Entity entity : results.asIterable()) {
+        // Limits number of comments added to the page
+            if (i < quantity){ 
+                long id = entity.getKey().getId();
+                String message = (String) entity.getProperty("comment");
+                Date timestamp = (Date) entity.getProperty("date");
+                String user = (String) entity.getProperty("email");
+                comments.add(new Comment(id, message, timestamp, user));
+                i++;
+            } else {
+                break; // Exits for-loop once requested number of comments appear
+            }
+        }
+        
+        response.setContentType("application/json");
+        String json = new Gson().toJson(comments);
+        response.getWriter().println(json);
+    } else {
+        String urlToRedirectToAfterUserLogsIn = "/comment.html";
+        String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+        response.getWriter().println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>");
+    }
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
 
+    UserService userService = UserServiceFactory.getUserService();
+
+    // Only logged-in users can post
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/data");
+      return;
+    }
+
     String newComment = request.getParameter("new-comment");
     String quant = request.getParameter("quantity");
```

Comment:
Nit. I think changing the variable name from quantity to max_comments might be more readable. Also, you can use something like quantity_parameter instead of shortening it to quant. That was it's clear that this variable is taken from the parameters and different from the member quantity variable. Having two similarly named variable can cause confusion.